### PR TITLE
Brancmodel1

### PR DESCRIPTION
**Pull request by dakotacookson**
# Description

Added Customer Controler with all gets needed as well as put and post
added customer tests as well

## Type of change
feature updates/additions

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# Testing Instructions
git fetch --all
git checkout  brancmodel1
open in visual studio and run the program as well as run the tests

# Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added test instructions that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by Daniel-Brewer — APPROVED** [review]
Great job! Looks good.

**Review by ColleenWoolsey — APPROVED** [review]
The CustomerController checks out in Postman as do the Test Results - Data was as expected for Get, Get single, Post and Put in both

File: BangazonAPI/Controllers/CustomerController.cs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        public CustomerController(IConfiguration config)
+        {
+            _config = config;
+        }
+        public SqlConnection Connection => new SqlConnection(_config.GetConnectionString("DefaultConnection"));
+        public async Task<IActionResult> Get()
+        {
+            using (SqlConnection conn = Connection)
+            {
+                List<object> customerControlerlist = new List<object>();
```
Comment (askingalot) [inline BangazonAPI/Controllers/CustomerController.cs]: Why is this a `List<object>`?

File: BangazonAPI/Controllers/CustomerController.cs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                    reader.Close();
+
+                    return Ok(customer);
+                }
+            }
+        }
+
+
+        [HttpGet]
+        public IEnumerable<Customer> Get(string q)
```
Comment (askingalot) [inline BangazonAPI/Controllers/CustomerController.cs:121]: I would like to see
`public IEnumerable<Customer> Get(string q)`
and
`public async Task<IActionResult> Get()`
combined into one method that has an optional `q` query string parameter.

File: BangazonAPI/Controllers/CustomerController.cs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    public class CustomerController : ControllerBase
+    {
+        private readonly IConfiguration _config;
+
+        public CustomerController(IConfiguration config)
+        {
+            _config = config;
+        }
+        public SqlConnection Connection => new SqlConnection(_config.GetConnectionString("DefaultConnection"));
+        public async Task<IActionResult> Get()
```
Comment (askingalot) [inline BangazonAPI/Controllers/CustomerController.cs]: Where is the "Get a single Customer" method?

File: Class1.cs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+﻿using System;
```
Comment (askingalot) [inline Class1.cs:1]: this file should not be here.

File: TestBangazonAPI/BangazonAPITests.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    public class CustomerTests
+    {
+        [Fact]
+        public async Task GetCustomer_Success()
+        {
+
+            using (var client = new APIClientProvider().Client)
+            {
+                var response = await client.GetAsync("api/Customer");
+                response.EnsureSuccessStatusCode();
```
Comment (askingalot) [inline TestBangazonAPI/BangazonAPITests.cs:20]: Please verify that you got at least one `Customer` back from the API.

File: TestBangazonAPI/BangazonAPITests.cs [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+            }
+        }
+        [Fact]
+        public async Task GetCustomers_Success()
+        {
```
Comment (askingalot) [inline TestBangazonAPI/BangazonAPITests.cs:32]: Please verify that you got at the expected `Customer` back from the API.

File: TestBangazonAPI/BangazonAPITests.cs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                var response = await client.GetAsync("api/Customer/5");
+                response.EnsureSuccessStatusCode();
+
+            }
+        }
+
+        [Fact]
+        public async Task GetCustomersq_Success()
+        {
+
```
Comment (askingalot) [inline TestBangazonAPI/BangazonAPITests.cs:48]: Please verify that you got a customer back that matches the `q` query string parameter.

File: TestBangazonAPI/BangazonAPITests.cs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+                var response = await client.GetAsync("api/Customer?q=test");
+                response.EnsureSuccessStatusCode();
+
+            }
+        }
+
+        [Fact]
+        public async Task PostCustomer_Success()
+        {
+
```
Comment (askingalot) [inline TestBangazonAPI/BangazonAPITests.cs:64]: Please verify that you got the expected response back from the API

File: TestBangazonAPI/BangazonAPITests.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        private string GenerateCustomer()
+        {
+            var customer = new Customer
+            {
+                FirstName = "Marshal",
+                LastName = "Lee"
+            };
+            Assert.Equal("Marshal", customer.FirstName);
```
Comment (askingalot) [inline TestBangazonAPI/BangazonAPITests.cs]: These asserts are unnecessary.

File: TestBangazonAPI/BangazonAPITests.cs [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            {
+                var customer = GenerateCustomers();
+                var response = await client.PutAsync("api/Customer/4", new StringContent(customer , Encoding.UTF8, "application/json"));
+                response.EnsureSuccessStatusCode();
+
+            }
+        }
+
+
+        private string GenerateCustomers()
```
Comment (askingalot) [inline TestBangazonAPI/BangazonAPITests.cs]: Why do you have this method as well as another called `GenerateCustomer`?

**Review by askingalot — CHANGES_REQUESTED** [review]

File: BangazonAPI/Controllers/CustomerController.cs (hunk 1 above)
Comment (dakotacookson) [reply BangazonAPI/Controllers/CustomerController.cs]: the generic form of any given object for use with multiple objects

File: BangazonAPI/Controllers/CustomerController.cs (hunk 2 above)
Comment (dakotacookson) [reply BangazonAPI/Controllers/CustomerController.cs:121]: thats more of a refactoring step atm we are just trying to finish this sprint

File: BangazonAPI/Controllers/CustomerController.cs (hunk 3 above)
Comment (dakotacookson) [reply BangazonAPI/Controllers/CustomerController.cs]: line 117...

File: Class1.cs (hunk 4 above)
Comment (dakotacookson) [reply Class1.cs:1]: my program disagrees with you taking it it breaks evrything

File: TestBangazonAPI/BangazonAPITests.cs (hunk 9 above)
Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs]: your literally contradicting yourself please explain?

File: TestBangazonAPI/BangazonAPITests.cs (hunk 10 above)
Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs]: well each method must be different one tests a single and one, a multiple

File: BangazonAPI/Controllers/CustomerController.cs (hunk 1 above)
Comment (dakotacookson) [reply BangazonAPI/Controllers/CustomerController.cs]: fixed

Comment (askingalot) [reply BangazonAPI/Controllers/CustomerController.cs]: But you know the true data type of each list element. Please use the correct type.

File: BangazonAPI/Controllers/CustomerController.cs (hunk 2 above)
Comment (askingalot) [reply BangazonAPI/Controllers/CustomerController.cs:121]: Please combine them into a single method.

File: Class1.cs (hunk 4 above)
Comment (askingalot) [reply Class1.cs:1]: Please remove it. If you get an error, figure out why and fix it.

File: BangazonAPI/Controllers/CustomerController.cs (hunk 3 above)
Comment (askingalot) [reply BangazonAPI/Controllers/CustomerController.cs]: Sorry I missed that.

File: TestBangazonAPI/BangazonAPITests.cs (hunk 10 above)
Comment (askingalot) [reply TestBangazonAPI/BangazonAPITests.cs]: That's not what this method does

File: TestBangazonAPI/BangazonAPITests.cs (hunk 9 above)
Comment (askingalot) [reply TestBangazonAPI/BangazonAPITests.cs]: This is not a test. You only have asserts inside a test.

File: BangazonAPI/Controllers/CustomerController.cs (hunk 2 above)
Comment (dakotacookson) [reply BangazonAPI/Controllers/CustomerController.cs:121]: why its unnecessary work that will keep us from hitting the end of our sprint successfully,  mind explaining your rationale?

File: Class1.cs (hunk 4 above)
Comment (dakotacookson) [reply Class1.cs:1]: fine

File: TestBangazonAPI/BangazonAPITests.cs (hunk 7 above)
Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs:48]: fixed

File: TestBangazonAPI/BangazonAPITests.cs (hunk 6 above)
Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs:32]: fixed

File: TestBangazonAPI/BangazonAPITests.cs (hunk 5 above)
Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs:20]: fixed

File: TestBangazonAPI/BangazonAPITests.cs (hunk 9 above)
Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs]: fixed

File: TestBangazonAPI/BangazonAPITests.cs (hunk 10 above)
Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs]: fine

Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs]: i do not understand your issue here

File: TestBangazonAPI/BangazonAPITests.cs (hunk 8 above)
Comment (dakotacookson) [reply TestBangazonAPI/BangazonAPITests.cs:64]: fixed

File: BangazonAPI/Controllers/CustomerController.cs (hunk 2 above)
Comment (dakotacookson) [reply BangazonAPI/Controllers/CustomerController.cs:121]: fixed

**Review by askingalot — APPROVED** [review]

**Pull request merged**